From 04634a9de457263e581b03b9518705cc156eecca Mon Sep 17 00:00:00 2001
From: Hoshino Lina <lina@lina.yt>
Date: Sat, 7 Mar 2026 19:52:05 +0900
Subject: [PATCH] linux-v4l2: Fix spurious fd closing

Make sure to use the invalid fd -1 when the output is not opened, and
only close valid fds. If fd 0 is closed, then this closes stdin. The
second time this happens, some other important fd will have become fd 0,
breaking something.

This causes random things to break (browser/CEF in reports, but really
it could be anything) as the wrong fds get closed.
---
 plugins/linux-v4l2/v4l2-output.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/plugins/linux-v4l2/v4l2-output.c b/plugins/linux-v4l2/v4l2-output.c
index 27abc5a5168d..dbb62777804d 100644
--- a/plugins/linux-v4l2/v4l2-output.c
+++ b/plugins/linux-v4l2/v4l2-output.c
@@ -27,7 +27,8 @@ static const char *virtualcam_name(void *unused)
 static void virtualcam_destroy(void *data)
 {
 	struct virtualcam_data *vcam = (struct virtualcam_data *)data;
-	close(vcam->device);
+	if (vcam->device >= 0)
+		close(vcam->device);
 	bfree(data);
 }
 
@@ -133,6 +134,7 @@ static void *virtualcam_create(obs_data_t *settings, obs_output_t *output)
 {
 	struct virtualcam_data *vcam = (struct virtualcam_data *)bzalloc(sizeof(*vcam));
 	vcam->output = output;
+	vcam->device = -1;
 
 	UNUSED_PARAMETER(settings);
 	return vcam;
@@ -256,6 +258,7 @@ static bool try_connect(void *data, const char *device)
 
 fail_close_device:
 	close(vcam->device);
+	vcam->device = -1;
 	return false;
 }
 
@@ -331,6 +334,7 @@ static void virtualcam_stop(void *data, uint64_t ts)
 	}
 
 	close(vcam->device);
+	vcam->device = -1;
 	blog(LOG_INFO, "Virtual camera stopped");
 
 	UNUSED_PARAMETER(ts);
-- 
2.53.0

